README.md
## Introduction

**BrentLab/callvariants** is a bioinformatics pipeline for general variant calling.
It runs Freebayes, TIDDIT and CNVpytor for SNP/INDEL and structural variant
calling.

This workflow has been developed with the following specific functionality
in mind:

- Checking the genotype of KN99alpha samples
  - This is performed by providing additional sequences to be appended to the
    genome prior to alignment in a per-smaple basis
- Processing *c. neoformans* samples for bulk segregant analysis
  - The Freebayes step can optionally be used to jointly call variants on
    groups which are identified in the input samplesheet

But there is no reason why it is limited to these applications.

The pipeline, overall, runs the following processes:

1. Prepare the Genome
    - Concatenate additional sequences provided in the input samplesheet,
    if there are any
    - Create indicies
        - [samtools faidx](http://www.htslib.org/doc/samtools-faidx.html)
        - [bwamem2 index](https://github.com/bwa-mem2/bwa-mem2)
        - [bwa index](https://github.com/lh3/bwa) -- this is for TIDDIT
    - Create sequence maps
        - [build](modules/local/build_intervals/main.nf) and
        [create](modules/local/create_intervals_bed/main.nf) intervals.
        Both of these are from [sarek](https://github.com/nf-core/sarek)
        - [GATK CreateSequenceDictionary](https://gatk.broadinstitute.org/hc/en-us/articles/360036712531-CreateSequenceDictionary-Picard-)
1. Read QC
    - [fastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
1. Align reads
    - [bwamem2](https://github.com/bwa-mem2/bwa-mem2)
    - [picard MarkDuplicates](https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard-)
    - [picard AddOrReplaceReadGroup](https://gatk.broadinstitute.org/hc/en-us/articles/360037226472-AddOrReplaceReadGroups)
    - [samtools](https://samtools.github.io/) index, sort, stats, flatstats, idxstats
1. Call Variants
    - [Freebayes](https://github.com/freebayes/freebayes)
    - [TIDDIT](https://github.com/SciLifeLab/TIDDIT)
    - [CNVpytor](https://github.com/abyzovlab/CNVpytor)
    - [snpEff](https://pcingola.github.io/SnpEff/)
    - [vcftools](https://vcftools.github.io/) for filtering
    - [bcftools](https://samtools.github.io/bcftools/bcftools.html) stats
1. Collect and present QC
    - [MultiQC](http://multiqc.info/)

## [Usage](docs/usage.md)

If you are new to Nextflow and nf-core, please refer to
[this page](https://nf-co.re/docs/usage/installation) on how to set-up
Nextflow. Make sure to
[test your setup](https://nf-co.re/docs/usage/introduction#how-to-run-a-pipeline)
with `-profile test` before running the workflow on actual data.
If you are running this test on WUSTL HTCF or RIS, use one of the built-in
[profiles](docs/usage.md#profile), either `htcf` or `ris`. If you are running
the test on a different host, then you may consider including one of the
dependency manager profiles, eg `singularity` or `docker`.

A test run for `ris`, for example, would look like this:

```bash
nextflow run BrentLab/callvariants -r main -profile ris,test
```
you will need to submit this appropriately, but no other input is necessary
to run the tests -- all input is taken care of by the `test` profile

For detailed instructions on running your own data, please see the
[**usage documentation**](docs/usage.md)

## [Output](docs/output.md)

For a description of the output directory, please see the
[**output documentation**](docs/output.md)

## Common problems

A pernicious error is the character that symbolizes a "new line" in a file. We
never see these characters, but they of course exist -- how else would the
computer know where the new line is?

Mac, Windows and Linux operating systems use different carriage return
characters, unfortunately. If you're using a cluster to process your data,
you need to make sure that the files are linux compliant. All of the files you
download from NCBI or fungiDB, for instance, will be, as will your fastq files
from the sequencer centers. But, if you create an additional fasta file, you
need to make sure that this hasn't been adulterated by Mac or Windows. I would
expect that snapgene would by default output a linux compliant fasta file. But
if you were to open the file in something like word, it would probably convert
the characters.

One tool you can use to ensure that your input files are linux compliant is
[dos2unix](https://dos2unix.sourceforge.io/). Using `dos2unix` looks like this:

```bash
dos2unix </path/to/file.<ext>>
```
and the file will be changed in-place.

## Credits

BrentLab/callvariants was originally written by Chase Mateusiak. It is based on
the BSA processing steps of Daniel Agustinhno.

## Contributions and Support

If you would like to contribute to this pipeline, please see the [contributing guidelines](.github/CONTRIBUTING.md).

## Citations

<!-- TODO after releasing a new version, update the doi url -->
If you use  BrentLab/callvariants for your analysis, please cite it using the following doi: [10.5281/zenodo.XXXXXX](https://doi.org/10.5281/zenodo.XXXXXX)

An extensive list of references for the tools used by the pipeline can be found in the [`CITATIONS.md`](CITATIONS.md) file.

This pipeline uses code and infrastructure developed and maintained by the [nf-core](https://nf-co.re) community, reused here under the [MIT license](https://github.com/nf-core/tools/blob/master/LICENSE).

> **The nf-core framework for community-curated bioinformatics pipelines.**
>
> Philip Ewels, Alexander Peltzer, Sven Fillinger, Harshil Patel, Johannes Alneberg, Andreas Wilm, Maxime Ulysse Garcia, Paolo Di Tommaso & Sven Nahnsen.
>
> _Nat Biotechnol._ 2020 Feb 13. doi: [10.1038/s41587-020-0439-x](https://dx.doi.org/10.1038/s41587-020-0439-x).


CHANGELOG.md
# BrentLab/callvariants: Changelog

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 1.3.0

Adding H99 configuration

## v1.0dev - [date]

Initial release of BrentLab/callvariants, created with the [nf-core](https://nf-co.re/) template.

### `Added`

### `Fixed`

### `Dependencies`

### `Deprecated`


assets/genome_files/h99_cnvpytor_haploid_genome_conf.py
import_reference_genomes = {
    "h99": {
        "name": "kn99_fungidb",
        "species": "cryptococcus neoformans",
        "chromosomes": OrderedDict([
            ("CP003820.1", (2291500, "S")),
            ("CP003821.1", (1621676, "S")),
            ("CP003822.1", (1574972, "S")),
            ("CP003823.1", (1084805, "S")),
            ("CP003824.1", (1814975, "S")),
            ("CP003825.1", (1422463, "S")),
            ("CP003826.1", (1399209, "S")),
            ("CP003827.1", (1398693, "S")),
            ("CP003828.1", (1186813, "S")),
            ("CP003829.1", (1059962, "S")),
            ("CP003830.1", (1562107, "S")),
            ("CP003831.1", (774060, "S")),
            ("CP003832.1", (756017, "S")),
            ("CP003833.1", (942472, "S")),
            ("CP003834.1", (24923, "M"))
        ]),
        "gc_file": "h99_gc_file.pytor"
    }
}


assets/genome_files/kn99_cnvpytor_haploid_genome_conf.py
import_reference_genomes = {
    "kn99": {
        "name": "kn99_fungidb",
        "species": "cryptococcus neoformans",
        "chromosomes": OrderedDict([
            ("CP022321.1", (2291500, "S")),
            ("CP022322.1", (1621676, "S")),
            ("CP022323.1", (1574972, "S")),
            ("CP022324.1", (1084805, "S")),
            ("CP022325.1", (1814975, "S")),
            ("CP022326.1", (1422463, "S")),
            ("CP022327.1", (1399209, "S")),
            ("CP022328.1", (1398693, "S")),
            ("CP022329.1", (1186813, "S")),
            ("CP022330.1", (1059962, "S")),
            ("CP022331.1", (1562107, "S")),
            ("CP022332.1", (774060, "S")),
            ("CP022333.1", (756017, "S")),
            ("CP022334.1", (942472, "S")),
            ("CP022335.1", (24923, "M")),
            ("NAT", (1650, "S")),
            ("G418", (2565, "S"))
        ]),
        "gc_file": "kn99_with_markers_gc.pytor"
    }
}
